### PR TITLE
Use one raft node to manage three stores

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: I64a363a46b806355cd844e655d51062b4cac2706

### CODE REVIEW

**Review by houzhizhen — DISMISSED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 1]
```diff
@@ -59,61 +58,71 @@
...
-        NodeOptions nodeOptions = context.nodeOptions(storePath);
+    private Node initRaftNode() throws IOException {
+        NodeOptions nodeOptions = this.context.nodeOptions();
         nodeOptions.setFsm(this.stateMachine);
 
-        // TODO: When support sharding, groupId needs to be bound to shard Id
-        String groupId = storePath;
-        PeerId endpoint = context.endpoint();
-
+        String groupId = this.context.group();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:119]: should keep comment "// TODO: When support sharding, groupId needs to be bound to shard Id"?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 2]
```diff
@@ -176,7 +186,7 @@ protected void waitLeaderElected(int timeout) {
                 } catch (InterruptedException e) {
                     throw new BackendException(
                               "Wait raft group '%s' election error",
-                              e, this.group(), "election");
+                              e, group, "election");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:189]: remove "election"

Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:189]: improve message: "Interrupted while waiting raft group '%s' election"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 3]
```diff
@@ -92,31 +108,69 @@ public RaftSharedContext(HugeGraphParams params) {
...
+        }
+    }
+
+    public BackendStore originStore(StoreType storeType) {
+        if (storeType == StoreType.SCHEMA) {
+            return this.stores.get(this.schemaStoreName).originStore();
+        } else if (storeType == StoreType.GRAPH) {
+            return this.stores.get(this.graphStoreName).originStore();
+        } else {
+            return this.stores.get(this.systemStoreName).originStore();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:164]: assert storeType == StoreType.SYSTEM

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java [hunk 4]
```diff
@@ -44,16 +53,18 @@ public StoreAction action() {
...
 
     public static StoreCommand fromBytes(byte[] bytes) {
-        StoreAction action = StoreAction.fromCode(bytes[0]);
-        byte[] data = new byte[bytes.length - 1];
-        System.arraycopy(bytes, 1, data, 0, bytes.length - 1);
-        return new StoreCommand(action, data);
+        StoreType type = StoreType.valueOf(bytes[0]);
+        StoreAction action = StoreAction.valueOf(bytes[1]);
+        byte[] data = new byte[bytes.length - 2];
+        System.arraycopy(bytes, 2, data, 0, bytes.length - 2);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java:67]: don't copy, just wrap by bytebuffer with offset

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java [hunk 5]
```diff
@@ -44,16 +53,18 @@ public StoreAction action() {
     }
 
     public byte[] toBytes() {
-        byte[] bytes = new byte[1 + this.data.length];
-        bytes[0] = this.action.code();
-        System.arraycopy(this.data, 0, bytes, 1, this.data.length);
+        byte[] bytes = new byte[1 + 1 + this.data.length];
+        bytes[0] = (byte) this.type.getNumber();
+        bytes[1] = (byte) this.action.getNumber();
+        System.arraycopy(this.data, 0, bytes, 2, this.data.length);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java:59]: reserve two bytes when allocate this.data

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 6]
```diff
@@ -92,31 +108,69 @@ public RaftSharedContext(HugeGraphParams params) {
...
+            return StoreType.GRAPH;
+        } else {
+            assert this.systemStoreName.equals(store);
+            return StoreType.SYSTEM;
+        }
+    }
+
+    public BackendStore originStore(StoreType storeType) {
+        if (storeType == StoreType.SCHEMA) {
+            return this.stores.get(this.schemaStoreName).originStore();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:160]: use array instead stores[StoreType.SCHEMA]

File: hugegraph-core/src/main/resources/raft.proto [hunk 7]
```diff
@@ -5,6 +5,12 @@ package com.baidu.hugegraph.backend.store.raft;
 option java_package="com.baidu.hugegraph.backend.store.raft";
 option java_outer_classname = "RaftRequests";
 
+enum StoreType {
+    SCHEMA = 1;
+    GRAPH = 2;
+    SYSTEM = 3;
```
Comment (javeme) [inline hugegraph-core/src/main/resources/raft.proto:11]: add SIZE

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 8]
```diff
@@ -287,6 +287,7 @@ public void setResponse(StoreCommandResponse resp) {
                     LOG.debug("StoreCommandResponse status error");
                     Status status = new Status(RaftError.UNKNOWN,
                                                "fowared request failed");
+                    System.out.println("失败的请求：" + request);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:290]: remove it

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java [hunk 9]
```diff
@@ -43,7 +43,11 @@
 
     public static byte[] writeMutations(List<BackendMutation> mutations) {
         int estimateSize = mutations.size() * MUTATION_SIZE;
-        BytesBuffer buffer = BytesBuffer.allocate(1 + estimateSize);
+        // The first two bytes are reserved for StoreType and StoreAction
+        BytesBuffer buffer = BytesBuffer.allocate(2 + 4 + estimateSize);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java:47]: define StoreCommand.HEADER_SIZE

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java [hunk 10]
```diff
@@ -43,7 +43,11 @@
...
+        buffer.write((byte) 0);
+        buffer.write((byte) 0);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java:49]: add static method StoreCommand.writeHeader(buffer)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java [hunk 11]
```diff
@@ -128,7 +132,11 @@ public static BackendMutation readMutation(BytesBuffer buffer) {
     }
 
     public static byte[] writeIncrCounter(IncrCounter incrCounter) {
-        BytesBuffer buffer = BytesBuffer.allocate(1 + BytesBuffer.LONG_LEN);
+        // The first two bytes are reserved for StoreType and StoreAction
+        BytesBuffer buffer = BytesBuffer.allocate(2 + 1 + BytesBuffer.LONG_LEN);
+        buffer.write((byte) 0);
+        buffer.write((byte) 0);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java:138]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 12]
```diff
@@ -93,31 +108,66 @@ public RaftSharedContext(HugeGraphParams params) {
...
+    public RaftNode node() {
+        return this.raftNode;
+    }
+
+    public String group() {
+        return DEFAULT_GROUP;
+    }
+
+    public void addStore(StoreType type, RaftBackendStore store) {
+        this.stores[type.ordinal()] = store;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:144]: also use getNumber()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 13]
```diff
@@ -93,31 +108,66 @@ public RaftSharedContext(HugeGraphParams params) {
...
-        if (!this.nodes.containsKey(group)) {
-            synchronized (this.nodes) {
-                if (!this.nodes.containsKey(group)) {
-                    LOG.info("Initing raft node for '{}'", group);
-                    RaftNode node = new RaftNode(group, store, this);
-                    this.nodes.put(group, node);
-                }
-            }
+    public BackendStore originStore(StoreType storeType) {
+        return this.stores[storeType.getNumber()].originStore();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:159]: check store is not null

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java [hunk 14]
```diff
@@ -19,20 +19,33 @@
...
     private final StoreAction action;
     private final byte[] data;
 
-    public StoreCommand(StoreAction action) {
-        this(action, EMPTY);
+    public StoreCommand(StoreType type, StoreAction action, byte[] data) {
+        this.type = type;
+        this.action = action;
+        if (data == null) {
+            this.data = new byte[2];
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java:38]: use HEADER_SIZE here

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java [hunk 15]
```diff
@@ -43,17 +56,20 @@ public StoreAction action() {
...
-        byte[] bytes = new byte[1 + this.data.length];
-        bytes[0] = this.action.code();
-        System.arraycopy(this.data, 0, bytes, 1, this.data.length);
+    public static void writeHeader(BytesBuffer buffer) {
+        buffer.write((byte) 0);
+        buffer.write((byte) 0);
+    }
+
+    public static byte[] wrap(byte value) {
+        byte[] bytes = new byte[3];
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommand.java:65]: use HEADER_SIZE here

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java (hunk 9 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java:47]: use HEADER_SIZE + 4 + estimateSize

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java [hunk 16]
```diff
@@ -128,7 +131,10 @@ public static BackendMutation readMutation(BytesBuffer buffer) {
     }
 
     public static byte[] writeIncrCounter(IncrCounter incrCounter) {
-        BytesBuffer buffer = BytesBuffer.allocate(1 + BytesBuffer.LONG_LEN);
+        // The first two bytes are reserved for StoreType and StoreAction
+        BytesBuffer buffer = BytesBuffer.allocate(2 + 1 + BytesBuffer.LONG_LEN);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreSerializer.java:135]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommandRequestProcessor.java [hunk 17]
```diff
@@ -75,9 +68,9 @@ public String interest() {
     }
 
     private StoreCommand parseStoreCommand(StoreCommandRequest request) {
-        byte actionByte = (byte) request.getAction().getNumber();
-        StoreAction action = StoreAction.fromCode(actionByte);
+        StoreType type = request.getType();
+        StoreAction action = request.getAction();
         byte[] data = request.getData().toByteArray();
-        return new StoreCommand(action, data);
+        return new StoreCommand(type, action, data);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommandRequestProcessor.java:74]: if request.getData() returns ByteBuffer, just wrap it with BytesBuffer then pass into  StoreCommand

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommandRequestProcessor.java (hunk 17 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreCommandRequestProcessor.java:74]: no, it return com.google.protobuf.ByteString

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java [hunk 18]
```diff
@@ -156,7 +157,10 @@ public StoreType storeType(String store) {
     }
 
     public BackendStore originStore(StoreType storeType) {
-        return this.stores[storeType.getNumber()].originStore();
+        RaftBackendStore raftStore = this.stores[storeType.getNumber()];
+        E.checkState(raftStore != null,
+                     "The raft store type %s shouldn't be null");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedContext.java:162]: missing parameter, and improve "The raft store of type %s shouldn't be null"

**Review by javeme — APPROVED** [review]

**Review by houzhizhen — APPROVED** [review]

**Review by zhoney — APPROVED** [review]

**Pull request merged**